recorder.cpp

```diff
@@ -170,44 +196,98 @@ struct RecorderImplementation : public Recorder {
             workspace.jFrame["cameraParameters"]["principalPointX"] = f.px;
             workspace.jFrame["cameraParameters"]["principalPointY"] = f.py;
         }
-
-        if (!videoOutputPrefix.empty() && f.frameData != nullptr) {
-            if (!videoWriters.count(f.cameraInd)) {
-                videoWriters[f.cameraInd] = VideoWriter::build(videoOutputPrefix, f.cameraInd, fps, *f.frameData);
-            }
-            videoWriters.at(f.cameraInd)->write(*f.frameData);
-        }
     }
 
-    void addFrame(const FrameData &f) final {
-        setFrame(f);
-        workspace.jFrame["number"] = frameNumberGroup;
-        workspace.jFrameGroup["time"] = f.t;
-        workspace.jFrameGroup["number"] = frameNumberGroup;
-        workspace.jFrameGroup["frames"] = {};
-        workspace.jFrameGroup["frames"].push_back(workspace.jFrame);
-        output << workspace.jFrameGroup.dump() << std::endl;
-        frameNumberGroup++;
+    void frameDrop(double time) {
+        jsonlProcessor->enqueue([this, time]() {
+            workspace.jFrameDrop["time"] = time;
+            output << workspace.jFrameDrop.dump() << std::endl;
+        });
     }
 
-    void addFrameGroup(double t, const std::vector<FrameData> &frames) final {
-        workspace.jFrameGroup["time"] = t;
-        workspace.jFrameGroup["number"] = frameNumberGroup;
-        workspace.jFrameGroup["frames"] = {};
+    #ifdef USE_OPENCV_VIDEO_RECORDING
+    bool allocateAndWriteVideo(const std::vector<FrameData> &frames) {
+        std::vector<std::shared_ptr<cv::Mat>> allocatedFrames;
```

Comment:
This could be a member variable (which would be just `.clear()`ed here) so nothing is allocated on heap on `push_back`